File: Views/Components/CoupleStatisticsView.swift
import SwiftUI
import Combine
import CoreLocation

struct CoupleStatisticsView: View {
    @EnvironmentObject var appState: AppState
    @StateObject private var journalService = JournalService.shared
    @StateObject private var categoryProgressService = CategoryProgressService.shared
    
    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            // Titre de la section
            HStack {
                Text("couple_statistics".localized)
                    .font(.system(size: 22, weight: .semibold))
                    .foregroundColor(.black)
                    .multilineTextAlignment(.center)
                    .padding(.horizontal, 20)
                
                Spacer()
            }
            
            // Grille de statistiques 2x3
            LazyVGrid(columns: [
                GridItem(.flexible(), spacing: 16),
                GridItem(.flexible(), spacing: 16)
            ], spacing: 16) {
                
                // Jours ensemble
                StatisticCardView(
                    title: "days_together".localized,
                    value: "\(daysTogetherCount)",
                    icon: "jours",
                    iconColor: Color(hex: "#feb5c8"),
                    backgroundColor: Color(hex: "#fedce3"),
                    textColor: Color(hex: "#db3556")
                )
                
                // Pourcentage de questions ouvertes
                StatisticCardView(
                    title: "questions_answered".localized,
                    value: "\(Int(questionsProgressPercentage))%",
                    icon: "qst",
                    iconColor: Color(hex: "#fed397"),
                    backgroundColor: Color(hex: "#fde9cf"),
                    textColor: Color(hex: "#ffa229")
                )
                
                // Villes visitées
                StatisticCardView(
                    title: "cities_visited".localized,
                    value: "\(citiesVisitedCount)",
                    icon: "ville",
                    iconColor: Color(hex: "#b0d6fe"),
                    backgroundColor: Color(hex: "#dbecfd"),
                    textColor: Color(hex: "#0a85ff")
                )
                
                // Pays visités
                StatisticCardView(
                    title: "countries_visited".localized,
                    value: "\(countriesVisitedCount)",
                    icon: "pays",
                    iconColor: Color(hex: "#d1b3ff"), // 🎯 Violet clair pour l'icône
                    backgroundColor: Color(hex: "#e8dcff"), // 🎯 Fond violet très clair
                    textColor: Color(hex: "#7c3aed") // 🎯 Violet plus foncé pour texte et valeur
                )
            }
            .padding(.horizontal, 20)
        }
        .onAppear {
            print("📊 CoupleStatisticsView: Vue apparue, calcul des statistiques")
            // Forcer le recalcul en accédant à la variable
            let _ = questionsProgressPercentage
            
            // Déclencher le géocodage rétroactif si nécessaire
            Task {
                await repairJournalEntriesGeocoding()
            }
        }
        .onReceive(categoryProgressService.$categoryProgress) { newProgress in
            print("📊 CoupleStatisticsView: Progression des catégories mise à jour: \(newProgress)")
            print("📊 CoupleStatisticsView: Recalcul du pourcentage...")
            // Forcer le recalcul
            let _ = questionsProgressPercentage
        }
    }
    
    // MARK: - Computed Properties
    
    /// Nombre de jours ensemble basé sur la date de début de relation
    private var daysTogetherCount: Int {
        guard let relationshipStartDate = appState.currentUser?.relationshipStartDate else {
            return 0
        }
        
        let calendar = Calendar.current
        let dayComponents = calendar.dateComponents([.day], from: relationshipStartDate, to: Date())
        return max(dayComponents.day ?? 0, 0)
    }
    
    /// Pourcentage de progression total sur toutes les questions
    private var questionsProgressPercentage: Double {
        let categories = QuestionCategory.categories
        var totalQuestions = 0
        var totalProgress = 0
        
        for category in categories {
            let questions = getQuestionsForCategory(category.id)
            // ⚠️ PROBLÈME POTENTIEL: On utilisait category.title au lieu de category.id
            let currentIndex = categoryProgressService.getCurrentIndex(for: category.id)
            
            totalQuestions += questions.count
            totalProgress += min(currentIndex + 1, questions.count) // +1 car l'index commence à 0
        }
        
        guard totalQuestions > 0 else { 
            return 0.0 
        }
        
        let percentage = (Double(totalProgress) / Double(totalQuestions)) * 100.0
        return percentage
    }
    
    /// Nombre de villes uniques visitées basé sur les entrées de journal
    private var citiesVisitedCount: Int {
        let uniqueCities = Set(journalService.entries.compactMap { entry in
            entry.location?.city?.trimmingCharacters(in: .whitespacesAndNewlines)
        }.filter { !$0.isEmpty })
        
        return uniqueCities.count
    }
    
    /// Nombre de pays uniques visités basé sur les entrées de journal
    private var countriesVisitedCount: Int {
        let uniqueCountries = Set(journalService.entries.compactMap { entry in
            entry.location?.country?.trimmingCharacters(in: .whitespacesAndNewlines)
        }.filter { !$0.isEmpty })
        
        return uniqueCountries.count
    }

    
    // MARK: - Helper Methods
    
    /// Récupère les questions pour une catégorie donnée via les données statiques
    private func getQuestionsForCategory(_ categoryId: String) -> [Question] {
        // Pour l'instant, utiliser les données statiques directement
        // TODO: Intégrer avec QuestionCacheManager quand la hiérarchie EnvironmentObject sera correcte
        return getQuestionsSampleForCategory(categoryId)
    }
    
    /// Charge les questions pour une catégorie via le nouveau QuestionDataManager
    private func getQuestionsSampleForCategory(_ categoryId: String) -> [Question] {
        return QuestionDataManager.shared.loadQuestions(for: categoryId)
    }
    
    /// Répare les entrées de journal existantes qui n'ont pas d'informations de ville/pays
    private func repairJournalEntriesGeocoding() async {
        let entriesToRepair = journalService.entries.filter { entry in
            // Entrées qui ont une localisation (coordonnées) mais pas de ville/pays
            guard let location = entry.location else { return false }
            let needsRepair = location.city == nil || location.country == nil || 
                             location.city?.isEmpty == true || location.country?.isEmpty == true ||
                             location.city == "unknown_city".localized || location.country == "unknown_location".localized ||
                             location.city == "Città sconosciuta" || location.country == "Sconosciuto"
            return needsRepair
        }
        
        guard !entriesToRepair.isEmpty else { return }
        
        // Limiter à 3 réparations par session pour éviter de surcharger l'API
        let limitedEntries = Array(entriesToRepair.prefix(3))
        
        for entry in limitedEntries {
            await repairSingleEntryGeocoding(entry)
        }
    }
    
    /// Répare une seule entrée de journal avec géocodage inversé
    private func repairSingleEntryGeocoding(_ entry: JournalEntry) async {
        guard let location = entry.location else { return }
        
        do {
            let clLocation = CLLocation(latitude: location.latitude, longitude: location.longitude)
            let geocoder = CLGeocoder()
            
            let placemarks = try await geocoder.reverseGeocodeLocation(clLocation)
            
            if let placemark = placemarks.first {
                let repairedLocation = JournalLocation(
                    coordinate: location.coordinate,
                    address: location.address ?? placemark.name,
                    city: placemark.locality,
                    country: placemark.country
                )
                
                // Créer une entrée mise à jour
                var updatedEntry = entry
                updatedEntry.location = repairedLocation
                updatedEntry.updatedAt = Date()
                
                // Sauvegarder via JournalService
                try await journalService.updateEntry(updatedEntry)
            }
        } catch {
            // Géocodage échoué, continuer silencieusement
        }
    }

}

// MARK: - Carte de statistique individuelle

struct StatisticCardView: View {
    let title: String
    let value: String
    let icon: String
    let iconColor: Color
    let backgroundColor: Color
    let textColor: Color
    
    var body: some View {
        VStack(spacing: 0) {
            // Ligne du haut : Icône à droite
            HStack {
                Spacer()
                
                // Icône en haut à droite
                Image(icon)
                    .resizable()
                    .aspectRatio(contentMode: .fit)
                    .frame(width: 40, height: 40)
                    .foregroundColor(iconColor)
            }
            
            Spacer()
            
            // Ligne du bas : Valeur + Titre à gauche
            HStack {
                VStack(alignment: .leading, spacing: 4) {
                    // Valeur principale
                    Text(value)
                        .font(.system(size: 32, weight: .bold))
                        .foregroundColor(textColor)
                        .minimumScaleFactor(0.7)
                        .lineLimit(1)
                    
                    // Titre
                    Text(title)
                        .font(.system(size: 14, weight: .medium))
                        .foregroundColor(textColor)
                        .multilineTextAlignment(.leading)
                        .lineLimit(2)
                        .fixedSize(horizontal: false, vertical: true)
                }
                
                Spacer()
            }
        }
        .frame(maxWidth: .infinity)
        .frame(height: 140)
        .padding(16)
        .background(
            RoundedRectangle(cornerRadius: 16)
                .fill(backgroundColor)
                .shadow(color: Color.black.opacity(0.05), radius: 8, x: 0, y: 2)
        )
    }
}

#Preview {
    CoupleStatisticsView()
        .environmentObject(AppState())
} 

File: Views/Components/CoupleStatisticsView.kt
// CoupleStatisticsScreen.kt
package com.love2loveapp.features.couple.stats

import android.app.Application
import android.location.Geocoder
import android.util.Log
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.love2loveapp.R
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import java.util.Locale
import kotlin.math.roundToInt

/**
 * ⚙️ Dépendances attendues côté Android (placeholders/interfaces) :
 * - AppState.shared.currentUser?.relationshipStartDate: java.util.Date? ou LocalDate?
 * - JournalService.shared.entries: List<JournalEntry>
 * - JournalService.shared.updateEntry(entry)
 * - CategoryProgressService.shared.getCurrentIndex(categoryId: String): Int
 * - QuestionDataManager.shared.loadQuestions(categoryId: String): List<Question>
 * - QuestionCategory.categories: List<QuestionCategory> avec .id
 *
 * Adapte les imports/DI (Hilt, Koin, etc.) à ton projet.
 */

// --- Modèles minimalistes pour aider l’intégration (adapte/branche sur tes vrais types) ---
data class AppUser(val relationshipStartDate: java.util.Date?)
data class JournalLocation(
    val latitude: Double,
    val longitude: Double,
    val address: String?,
    val city: String?,
    val country: String?
)
data class JournalEntry(
    val id: String,
    var location: JournalLocation?,
    var updatedAt: java.util.Date?
)
data class Question(val id: String)
data class QuestionCategory(val id: String, val title: String)

// --- Singletons placeholders (remplace par tes services réels) ---
object AppState {
    var currentUser: AppUser? = null
    // Fournis un accès partagé si nécessaire
    val shared = this
}

object JournalService {
    // Dans ton app réelle, expose en Flow/LiveData. Ici on simplifie.
    var entries: List<JournalEntry> = emptyList()
    suspend fun updateEntry(updated: JournalEntry) { /* write to storage */ }
    val shared = this
}

object CategoryProgressService {
    fun getCurrentIndex(categoryId: String): Int = 0
    val shared = this
}

object QuestionDataManager {
    fun loadQuestions(categoryId: String): List<Question> = emptyList()
    val shared = this
}

object QuestionCategoryRepository {
    val categories: List<QuestionCategory> = emptyList()
}

// --- UI State ---
data class StatsUiState(
    val daysTogether: Int = 0,
    val questionsProgressPercent: Int = 0,
    val citiesVisited: Int = 0,
    val countriesVisited: Int = 0
)

// --- ViewModel ---
class CoupleStatisticsViewModel(
    application: Application
) : AndroidViewModel(application) {

    private val appState = AppState.shared
    private val journalService = JournalService.shared
    private val categoryProgressService = CategoryProgressService.shared
    private val questionDataManager = QuestionDataManager.shared

    private val _uiState = MutableStateFlow(StatsUiState())
    val uiState: StateFlow<StatsUiState> = _uiState

    // Libellés localisés nécessaires côté VM (équivalent à context.getString(...))
    private val unknownCity: String =
        application.getString(R.string.unknown_city)
    private val unknownLocation: String =
        application.getString(R.string.unknown_location)

    fun onAppear() {
        Log.d("📊 CoupleStatistics", "Vue apparue, calcul des statistiques")
        recomputeAll()
        // Réparation géocodage (max 3)
        viewModelScope.launch { repairJournalEntriesGeocoding() }
    }

    fun onCategoryProgressChanged() {
        Log.d("📊 CoupleStatistics", "Progression catégories mise à jour → recalcul")
        recomputeQuestionsProgressOnly()
    }

    private fun recomputeAll() {
        val days = computeDaysTogether()
        val qPercent = computeQuestionsProgressPercent()
        val (cities, countries) = computeVisitedPlacesCounts()
        _uiState.value = StatsUiState(
            daysTogether = days,
            questionsProgressPercent = qPercent,
            citiesVisited = cities,
            countriesVisited = countries
        )
    }

    private fun recomputeQuestionsProgressOnly() {
        val current = _uiState.value
        _uiState.value = current.copy(
            questionsProgressPercent = computeQuestionsProgressPercent()
        )
    }

    private fun computeDaysTogether(): Int {
        val start = appState.currentUser?.relationshipStartDate ?: return 0
        val startLocal = start.toInstant().atZone(ZoneId.systemDefault()).toLocalDate()
        val today = LocalDate.now()
        val days = ChronoUnit.DAYS.between(startLocal, today).toInt()
        return days.coerceAtLeast(0)
    }

    private fun computeQuestionsProgressPercent(): Int {
        val categories = if (QuestionCategoryRepository.categories.isNotEmpty())
            QuestionCategoryRepository.categories
        else
            emptyList()

        var totalQuestions = 0
        var totalProgress = 0

        for (category in categories) {
            val questions = questionDataManager.loadQuestions(category.id)
            val currentIndex = categoryProgressService.getCurrentIndex(category.id)
            totalQuestions += questions.size
            totalProgress += minOf(currentIndex + 1, questions.size) // index base 0 → +1
        }

        if (totalQuestions == 0) return 0
        return ((totalProgress.toDouble() / totalQuestions.toDouble()) * 100.0)
            .roundToInt()
            .coerceIn(0, 100)
    }

    private fun computeVisitedPlacesCounts(): Pair<Int, Int> {
        val entries = journalService.entries

        val uniqueCities = entries.mapNotNull { it.location?.city?.trim() }
            .filter { it.isNotEmpty() && it != unknownCity }
            .toSet()

        val uniqueCountries = entries.mapNotNull { it.location?.country?.trim() }
            .filter { it.isNotEmpty() && it != unknownLocation }
            .toSet()

        return uniqueCities.size to uniqueCountries.size
    }

    private suspend fun repairJournalEntriesGeocoding() {
        val entriesToRepair = journalService.entries.filter { entry ->
            val loc = entry.location
            if (loc == null) return@filter false
            val cityMissing = loc.city.isNullOrBlank() || loc.city == unknownCity
            val countryMissing = loc.country.isNullOrBlank() || loc.country == unknownLocation
            (cityMissing || countryMissing)
        }.take(3)

        if (entriesToRepair.isEmpty()) return

        @Suppress("DEPRECATION")
        val geocoder = Geocoder(getApplication(), Locale.getDefault())

        viewModelScope.launch(Dispatchers.IO) {
            entriesToRepair.forEach { entry ->
                runCatching {
                    val loc = entry.location ?: return@runCatching
                    val results = geocoder.getFromLocation(loc.latitude, loc.longitude, 1)
                    if (!results.isNullOrEmpty()) {
                        val a = results[0]
                        val repaired = loc.copy(
                            address = loc.address ?: a.featureName,
                            city = a.locality ?: a.subAdminArea ?: loc.city ?: unknownCity,
                            country = a.countryName ?: loc.country ?: unknownLocation
                        )
                        val updated = entry.copy(
                            location = repaired,
                            updatedAt = java.util.Date()
                        )
                        journalService.updateEntry(updated)
                    }
                }.onFailure {
                    // Géocodage échoué → on continue silencieusement
                }
            }
        }
    }
}

// --- UI Composables ---
@Composable
fun CoupleStatisticsScreen(
    viewModel: CoupleStatisticsViewModel,
) {
    val uiState by viewModel.uiState.collectAsState()
    val context = LocalContext.current

    // 🎯 i18n via strings.xml
    val title = stringResource(id = R.string.couple_statistics)
    val daysTitle = stringResource(id = R.string.days_together)
    val questionsAnsweredTitle = stringResource(id = R.string.questions_answered)
    val citiesVisitedTitle = stringResource(id = R.string.cities_visited)
    val countriesVisitedTitle = stringResource(id = R.string.countries_visited)

    // Déclencheurs "onAppear" & "onReceive"
    LaunchedEffect(Unit) {
        viewModel.onAppear()
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(top = 16.dp, bottom = 8.dp)
    ) {
        // Titre
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = title,
                style = MaterialTheme.typography.titleLarge.copy(
                    fontSize = MaterialTheme.typography.titleLarge.fontSize,
                    fontWeight = FontWeight.SemiBold,
                    color = Color.Black
                ),
                modifier = Modifier.padding(horizontal = 20.dp)
            )
            Spacer(modifier = Modifier.weight(1f))
        }

        Spacer(modifier = Modifier.height(16.dp))

        // Grille 2x2
        val cards = listOf(
            StatCardData(
                title = daysTitle,
                value = uiState.daysTogether.toString(),
                iconRes = R.drawable.jours,              // 🖼️ place ton asset dans res/drawable
                iconColor = colorFromHex("#feb5c8"),
                background = colorFromHex("#fedce3"),
                textColor = colorFromHex("#db3556")
            ),
            StatCardData(
                title = questionsAnsweredTitle,
                value = "${uiState.questionsProgressPercent}%",
                iconRes = R.drawable.qst,                // 🖼️
                iconColor = colorFromHex("#fed397"),
                background = colorFromHex("#fde9cf"),
                textColor = colorFromHex("#ffa229")
            ),
            StatCardData(
                title = citiesVisitedTitle,
                value = uiState.citiesVisited.toString(),
                iconRes = R.drawable.ville,              // 🖼️
                iconColor = colorFromHex("#b0d6fe"),
                background = colorFromHex("#dbecfd"),
                textColor = colorFromHex("#0a85ff")
            ),
            StatCardData(
                title = countriesVisitedTitle,
                value = uiState.countriesVisited.toString(),
                iconRes = R.drawable.pays,               // 🖼️
                iconColor = colorFromHex("#d1b3ff"),
                background = colorFromHex("#e8dcff"),
                textColor = colorFromHex("#7c3aed")
            )
        )

        LazyVerticalGrid(
            columns = GridCells.Fixed(2),
            contentPadding = PaddingValues(horizontal = 20.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
            horizontalArrangement = Arrangement.spacedBy(16.dp),
            modifier = Modifier.fillMaxWidth()
        ) {
            items(cards) { card ->
                StatisticCard(
                    title = card.title,
                    value = card.value,
                    iconRes = card.iconRes,
                    iconTint = card.iconColor,
                    background = card.background,
                    textColor = card.textColor
                )
            }
        }
    }
}

data class StatCardData(
    val title: String,
    val value: String,
    val iconRes: Int,
    val iconColor: Color,
    val background: Color,
    val textColor: Color
)

@Composable
fun StatisticCard(
    title: String,
    value: String,
    iconRes: Int,
    iconTint: Color,
    background: Color,
    textColor: Color
) {
    Box(
        modifier = Modifier
            .height(140.dp)
            .fillMaxWidth()
            .background(background, shape = androidx.compose.foundation.shape.RoundedCornerShape(16.dp))
            .padding(16.dp)
    ) {
        // Icône top-right
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.End
        ) {
            Image(
                painter = painterResource(id = iconRes),
                contentDescription = null,
                modifier = Modifier.size(40.dp)
            )
            // Si tu veux teinter un VectorDrawable, remplace Image par Icon et passe tint = iconTint
        }

        // Valeur + titre bottom-left
        Column(
            modifier = Modifier
                .fillMaxHeight()
                .align(Alignment.BottomStart)
        ) {
            Text(
                text = value,
                fontWeight = FontWeight.Bold,
                color = textColor,
                style = MaterialTheme.typography.headlineSmall,
                maxLines = 1
            )
            Spacer(modifier = Modifier.height(4.dp))
            Text(
                text = title,
                color = textColor,
                style = MaterialTheme.typography.bodyMedium,
                maxLines = 2
            )
        }
    }
}

// --- Utils ---
fun colorFromHex(hex: String): Color {
    val clean = hex.removePrefix("#")
    val parsed = clean.toLong(radix = 16)
    return when (clean.length) {
        6 -> Color(0xFF000000 or parsed)
        8 -> Color(parsed)
        else -> Color.Unspecified
    }
}
